core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -475,7 +475,7 @@ private[log] class Cleaner(val id: Int,
       if (writeBuffer.position > 0) {
         writeBuffer.flip()
         val retained = MemoryRecords.readableRecords(writeBuffer)
-        dest.append(firstOffset = retained.deepEntries.iterator.next().offset,
+        dest.append(firstOffset = retained.batches.iterator.next().baseOffset,
```

Comment:
We stated in `RecordBatch` that `baseOffset` remains the same after compaction. It would be good to specify the behaviour of other header fields like `lastOffset` and `maxTimestamp`. Furthermore, we should make sure to test the specified behaviour for these cases (i.e. compaction removes the first/last item in the batch, compaction removed the item with the max timestamp). Do we have these tests already?